shark_turbine/kernel/wave/wave.py

```diff
@@ -0,0 +1,91 @@
+from typing import Callable, Optional
+import inspect
+
+from ..compiler.ir import Context, Operation
+from ..lang import Grid
+from .._support.tracing import (
+    CapturedTrace,
+    CompiledContext,
+    KernelRegionGraph,
+    Launchable,
+)
+from .._support.nodes import *
+
+
+__all__ = ["wave"]
+
+
+def wave():
+    def decorator(f: Callable[[Any], Any]) -> "LaunchableWave":
+        return LaunchableWave(f.__name__, f)
+
+    return decorator
+
+
+class LaunchableWave(Launchable):
+    def __init__(
+        self,
+        name: str,
+        eager_function: Callable[[Any], Any],
+    ):
+        super().__init__(eager_function)
+
+        self.grid_type = Grid[None, None]
+        self._name = name
+        self._f = eager_function
+        self._sig = inspect.signature(eager_function)
+
+    def _trace(self, dump: bool = False) -> CapturedTrace:
+        region_graph = KernelRegionGraph()
+        with CompiledContext(region_graph, grid_type=self.grid_type) as context:
+            custom_ops: dict[str, Type[CustomNode]] = {
+                "construct_register_from_metadata": ConstructRegisterFromMetadataNode,
```

Comment:
Do we need to specify all new nodes here or can they be inferred from one single source of truth?

Reply:
I reworked how new operations are registered. The single source of truth is `shark_turbine/kernel/ops/wave_ops`, e.g. [here](
https://github.com/iree-org/iree-turbine/blob/ec5b437ed1922cb6d0aea1e121d6ddd86cc3f307/shark_turbine/kernel/ops/wave_ops.py#L182)

This generates the functions used to construct and trace a program. The current limitation is proper IDE support as these functions are not statically known. I intend to work on that next and explore using `.pyi` stubs for that.